File: study/Algorithm/Trie/Trie.py
import sys


class Node(object):
    def __init__(self):
        self.valid = False
        self.children = [-1] * 26


class Trie:

    def __init__(self):
        self.trie = []
        self.root = self.init()

    def init(self):
        x = Node()
        self.trie.append(x)
        return len(self.trie) - 1

    def add(self, node, s, idx):
        if idx == len(s):
            self.trie[node].valid = True
            return
        c = ord(s[idx]) - ord('a')
        if self.trie[node].children[c] == -1:
            nx = self.init()
            self.trie[node].children[c] = nx
        child = self.trie[node].children[c]
        self.add(child, s, idx + 1)

    def addString(self, s):
        self.add(self.root, s, 0)

    def search(self, node, s, idx):
        if node == -1:
            return False
        if idx == len(s):
            return self.trie[node].valid
        c = ord(s[idx]) - ord('a')
        child = self.trie[node].children[c]
        return self.search(child, s, idx + 1)

    def searchString(self, s):
        return self.search(self.root, s, 0)


trie = Trie()
n, m = map(int, sys.stdin.readline().split())
ans = 0
for i in range(n):
    trie.addString(sys.stdin.readline().rstrip())
for j in range(m):
    if trie.searchString(sys.stdin.readline().rstrip()):
        ans += 1

print(ans)

File: study/Algorithm/Trie/Trie.cpp
struct Trie {
    struct Node {
        int children[26]; // 소문자만 추가
        bool valid;

        Node() {
            for (int &i : children) {
                i = -1;
            }
            valid = false;// true 추가한, false 중간
        }
    };

    vector<Node> trie;
    int root;

    int init() {
        Node x;
        trie.push_back(x);
        return (int) trie.size() - 1; // 추가한 노드의 idx 리턴
    }

    Trie() {
        root = init();
    }

    // node 탐색하고 있는 노드의 인덱스 s 추가하려고 하는 문자열 s 의 인덱스를 추가하고 있는가
    void add(int node, string &s, int index) {
        if (index == s.size()) {
            trie[node].valid = true;
            return;
        }
        int c = s[index] - 'a';
        if (trie[node].children[c] == -1) {
            int next = init();
            trie[node].children[c] = next;
        }
        int child = trie[node].children[c];
        add(child, s, index + 1);

    }

    bool search(int node, string &s, int index) {
        if (node == -1)return false;
        if (index == s.length())return trie[node].valid;
        int c = s[index] - 'a';
        int child = trie[node].children[c];
        return search(child, s, index + 1);
    }
};